desc/integrals/interp_utils.py

```diff
@@ -2,17 +2,458 @@
+):
+    """Evaluate Fourier coefficients ``a`` at coordinates ``(xq0,xq1)``.
+
+    Parameters
+    ----------
+    xq0 : jnp.ndarray
+        Real query points of coordinate in ``domain0`` where interpolation is desired.
+        Shape must broadcast with shape ``np.delete(a.shape,axes)``.
+        The coordinates stored here must be the same coordinate enumerated
+        across axis ``min(axes)`` of the Fourier coefficients ``a``.
+    xq1 : jnp.ndarray
+        Real query points of coordinate in ``domain1`` where interpolation is desired.
+        Shape must broadcast with shape ``np.delete(a.shape,axes)``.
+        The coordinates stored here must be the same coordinate enumerated
+        across axis ``max(axes)`` of the Fourier coefficients ``a``.
+    a : jnp.ndarray
+        Shape (..., a.shape[-2], a.shape[-1]).
+        Fourier coefficients ``a=rfft2(f,axes=axes,norm="forward")``.
+    n0 : int
+        Spectral resolution of ``a`` for ``domain0``.
+    n1 : int
+        Spectral resolution of ``a`` for ``domain1``.
+    domain0 : tuple[float]
+        Domain of coordinate specified by ``xq0`` over which samples were taken.
+    domain1 : tuple[float]
+        Domain of coordinate specified by ``xq1`` over which samples were taken.
+    axes : tuple[int]
+        Axes along which to transform.
+
+    Returns
+    -------
+    fq : jnp.ndarray
+        Real function value at query points.
+
+    """
+    idx = np.argsort(axes)
+    d = (domain0, domain1)
+    n = (n0, n1)
+    d_fft, d_rfft = d[idx[0]], d[idx[1]]
+    n_fft, n_rfft = n[idx[0]], n[idx[1]]
+
+    # |a| << |xq|, so move a instead
+    a = (
+        jnp.moveaxis(a, source=axes, destination=(-2, -1))
+        .at[..., 0]
+        .divide(2.0)
+        .at[..., -1]
+        .divide(1.0 + ((n_rfft % 2) == 0))
+    )
+    n_fft = jnp.fft.fftfreq(n_fft, d=(d_fft[1] - d_fft[0]) / (2 * jnp.pi * n_fft))
+    n_rfft = jnp.fft.rfftfreq(n_rfft, d=(d_rfft[1] - d_rfft[0]) / (2 * jnp.pi * n_rfft))
+    xq = (xq0 - domain0[0], xq1 - domain1[0])
+    basis = jnp.exp(
+        1j
+        * (
+            (n_fft * xq[idx[0]][..., jnp.newaxis])[..., jnp.newaxis]
+            + (n_rfft * xq[idx[1]][..., jnp.newaxis])[..., jnp.newaxis, :]
+        )
+    )
+    return 2.0 * (basis * a).real.sum(axis=(-2, -1))
+
+
+def cheb_from_dct(a, axis=-1):
+    """Get discrete Chebyshev transform from discrete cosine transform.
+
+    Parameters
+    ----------
+    a : jnp.ndarray
+        Discrete cosine transform coefficients, e.g.
+        ``a=dct(f,type=2,axis=axis,norm="forward")``.
+        The discrete cosine transformation used by scipy is defined here:
+        https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.dct.html.
+    axis : int
+        Axis along which to transform.
+
+    Returns
+    -------
+    cheb : jnp.ndarray
+        Chebyshev coefficients along ``axis``.
+
+    """
+    return a.copy().at[Index.get(0, axis, a.ndim)].divide(2.0)
```

Comment:
In normal numpy, if we removed the `.copy`, then `PiecewiseChebyshevSeries.evaluate` will modify the`self.cheb` and subsequent calls of `evaluate` method would be incorrect.  In not-compiled JAX, it doesn't do in place operations, but if the compiled corde decides to reuse the buffer than I think we'd see the same issue. In could lead to bad bugs as the tests don't use JIT but optimization does.

Given that the goal of this PR is to improve performance, should I always work with the cosines instead of the chebyshev polynomials to avoid these copies?

Reply:
> the compiler should only do the operation in place if its safe to do so

How does it decide safe?

This
```python
return a.at[Index.get(0, axis, a.ndim)].divide(2.0)
```
is equivalent to
```python
a[Index.get(0, axis, a.ndim)] /= 2.0
return a
```

Reply:
those two are not necessarily equivalent, it depends on where the return value goes. 

ie, 
```
a = a.at[i].divide(2)
```
is equivalent to 
```
a[i] /= 2
```
but different from 
```
b = a.at[i].divide(2)
```

In the first case under jit it will happen in place, the second generally won't be (unless `a` is never referenced again in which case its safe to do it in place and reuse the buffer `a->b`)

My point is that it should always be safe to do `a.at[i].divide(2)` without the explicit copy, since semantically you can always reason about it as if its returning a new array, not an in place modification.

Reply:
Ok thanks for explanation.

In that case we should fix `desc.backend.put`. Typically we use that as a replacement for the `a.at` jax indexing above. Had I naively used that here, an explicit copy _would_ be needed. Do you agree?

Reply:
This is why for example there is a `copy` flag in `grid.replace_at_axis`

Reply:
I think we can just get rid of the special case of that for numpy arrays?

Reply:
but then ``data["X"] = put(data["Y"], 1, 0)`` modifies data["Y"]

Reply:
No in that case Y should *not* be modified. Only if you did data[Y] = put(data[Y], ...)

Reply:
No, Y _is_ modified on numpy but not on JAX.

import numpy as np

def put(arr, inds, vals):
    """Functional interface for array "fancy indexing".

    Provides a way to do arr[inds] = vals in a way that works with JAX.

    Parameters
    ----------
    arr : array-like
        Array to populate
    inds : array-like of int
        Indices to populate
    vals : array-like
        Values to insert

    Returns
    -------
    arr : array-like
        Input array with vals inserted at inds.

    """
    arr[inds] = vals
    return arr

x = np.arange(10)
y = x**2
x = put(y, 0, -100000000000)
print(y)

Reply:
The bugfix is to add an explicit copy to the numpy branch of `desc.backend.put`.

Reply:
Ah I was looking at the Jax backend where we also have a special case for numpy arrays (I guess so we don't cast to Jax without knowing?). Yeah we should fix it in both places

Reply:
I am taking your word that, on jax
```data["X"] = function_of_put(data["Y"], 1, 0)```
will not modify Y even when `function_of_put` is written like
```python
def function_of_put(x...):
    x = put(x ...)
    return x
```

Reply:
actually, please confirm this, because it seems like it shouldn't

Reply:
pretty sure in that case it would get modified, since thats basically equilvanent to 
```
y = y.at[i].set(....)
x = y
```

Reply:
everything in jax gets inlined, so the compiler doesn't see function boundaries